packages/xforms-engine/test/parse/xpath/dependency-analysis.test.ts

```diff
@@ -0,0 +1,337 @@
+
+			contextNodeset: null,
+			expression: 'foo/bar | bat',
+			expected: ['foo/bar', 'bat'],
+		},
+
+		{
+			description:
+				'Relative context; relative paths in expression; extracted and contextualized as relative to the input context',
+
+			contextNodeset: 'quux',
+			expression: 'foo/bar or bat',
+			expected: ['quux/foo/bar', 'quux/bat'],
+		},
+
+		{
+			description:
+				'Absolute context; relative paths in expression; extracted and contextualized as relative to the input context',
+
+			contextNodeset: '/quux',
+			expression: 'foo/bar or bat',
+			expected: ['/quux/foo/bar', '/quux/bat'],
+		},
+
+		{
+			description:
+				'Absolute context; relative paths with abbreviated self and parent steps; extracted, and abbreviated steps resolved, relative to the input context',
+
+			contextNodeset: '/data/grp',
+			expression: './foo/bar and ../quux',
+			expected: ['/data/grp/foo/bar', '/data/quux'],
+		},
+
+		{
+			description:
+				'Absolute context; current() FilterExpr paths (one with a subsequent abbreviated parent step); extracted, current() treated as analogous to self step, parent step resolved, finally relative to the input context',
+
+			contextNodeset: '/data/grp',
+			expression: 'current()/foo/bar and current()/../quux',
+			expected: ['/data/grp/foo/bar', '/data/quux'],
+		},
+
+		{
+			description: 'No context; instance() and relative path; extracted and returned and as-is',
+
+			contextNodeset: null,
+			expression: 'instance("foo")/bar | ./quux',
+			expected: ['instance("foo")/bar', './quux'],
+		},
+
+		{
+			description:
+				'Relative context; instance() and relative path; extracted, instance call returned as-is, relative paths resolved to context',
+
+			contextNodeset: 'sel1',
+			expression: 'instance("foo")/bar and (bat or ./quux)',
+			expected: ['instance("foo")/bar', 'sel1/bat', 'sel1/quux'],
+		},
+
+		{
+			description:
+				'Relative context; instance() and relative path; extracted, instance call returned as-is, relative paths resolved to context',
+
+			contextNodeset: 'sel1',
+			expression: 'instance("foo")/bar and (bat or ../quux)',
+			expected: ['instance("foo")/bar', 'sel1/bat', 'sel1/../quux'],
+		},
+
+		{
+			description:
+				'Absolute context; instance() and relative path; extracted, instance call returned as-is, relative paths resolved to context',
+
+			contextNodeset: '/data/sel1',
+			expression: 'instance("foo")/bar and (bat or ../quux)',
+			expected: ['instance("foo")/bar', '/data/sel1/bat', '/data/quux'],
+		},
+
+		{
+			description:
+				'Context is instance("id") descendant. Expression references another instance (resolved as-is); relative sibling and relative child (resolved from context); relative sibling and child with current() calls (likewise resolved from context); absolute reference (resolved as-is)',
+
+			contextNodeset: 'instance("foo")/bar',
+			expression:
+				'if(position() > 2, instance("quux")/zig, ../zag | ./zag | current()/zag | current()/../zag) or /beep//boop',
```

Comment:
`current()/zag` resolves to the same as `./zag` and same with the ../ pair. Is that intentional? Seems separating them would reduce the risk of this masking a future bug (seems very low given other test coverage but possible)

Reply:
I'm not sure whether you're asking:

1. Whether it's correct that each pair resolves to the same path expression
2. Whether, when multiple sub-expressions resolve to the same path, we should produce the same path for each sub-expression respectively versus the current set-like behavior

If the former, I think it's correct but feel free to clarify in case I'm missing something!

If the latter, I don't know what benefit we'd get from producing the same path multiple times. The set-like behavior is intentional: we're determining which dependencies are referenced in an arbitrary expression, within a particular context... and then using those resolved paths to look up `*Node`s in instance state, establishing reactive subscriptions to those nodes to determine when the expression should be re-evaluated. Producing multiple same-path strings would probably be pure overhead.

We might reconsider the set-like aspect in, say, a context like dev tools or other form analysis tooling, where it could be useful to match found sub-expressions up with their specific resolved paths. But even then we'd probably also want to produce the unresolved sub-expression in the same result (and possibly even character indexes or some other way to identify where in the overall expression the result is found).

Reply:
I don't mean to suggest any change to the code, just to the test so that each case is separate in the result set. It's not very important, it's really just to make reading the test a little easier. Here I wasn't confused reading the test like I was with the `//../foo` case above.